### ISSUE TITLE
rustdoc-json: Guard agains 2 commits making the same modification to `FORMAT_VERSION`

### ISSUE THREAD

**Issue by aDotInTheVoid** [rust-lang/rust#94591]
If merge A and merge B both try to increase the version from 10 to 11, git will helpfully merge both of them, resulting in the version 11 having both changes. If a nightly is released in the time between A and B are merged, then both will claim to have the save version, but with different features.

This almost happened with #94009 and #94150, but was prevented by manual noticing and intervension, which is not ideal.

We should figure out a way such that if 2 commits both attempt to merge the same increments, git will raise this as a merge conflict, so this situation wont occor again.

A way to do this would be to store in a comment the logical "latest feature"

Eg: PR Foo

```diff
-// Latest Feature: Initial release
-const FORMAT_VERSION = 1;
+// Latest Feature: Foo
+const FORMAT_VERSION = 2;
```

PR Bar:
```diff
-// Latest Feature: Initial release
-const FORMAT_VERSION = 1;
+// Latest Feature: Bar
+const FORMAT_VERSION = 2;
```

If Foo is merged, when Bar is attempted to be merged, the following conflict will occor

```rust
<<<<<<< HEAD
// Latest Feature: Foo
=======
// Latest Feature: Bar
>>>>>>> bar
const FORMAT_VERSION = 2;
```

This could work, but will have a problem if two features are given the same name, and then we're back to where we were. If someone knows some git-trickery to get around that, that would be great, but if not, I still thing doing this (or something to this effect) would be worthwhile.

cc @Dylan-DPC @CraftSpider 

@rustbot modify labels: +T-rustdoc +A-rustdoc-json

**Comment by Dylan-DPC** [comment]
cc @rust-lang/rustdoc

**Comment by GuillaumeGomez** [comment]
I'm not sure what the best solution would be here... Adding a tidy check on the version number wouldn't work in this case unless you rebased your PR. Maybe adding this comment would be better, not sure...

**Comment by est31** [comment]
You could make a `const FORMAT_CHANGELOG = ["Initial version", "Done foo", "Done bar", ...];`, and then require PRs that change the version to add to that list. Then you could have `const FORMAT_VERSION = FORMAT_CHANGELOG.len();`. Or you could alternatively do it with PR numbers, but those are not known upon filing.

**Comment by camelid** [comment]
Do we definitely _need_ a format version?

**Comment by aDotInTheVoid** [comment]
Anacdotaly it is used to detect when the json moves, which does happen quite a bit: 

https://github.com/awslabs/smithy-rs/blob/d823f61156577ab42590709627906d1dc35a5f49/tools/api-linter/src/cargo.rs#L78-L96

Iff and when where talking about stabilization, we should reconsider, but I think it works well for now

**Comment by aDotInTheVoid** [comment]
cc @hkmatsumoto @disanti @Urgau @Maan2003 @akiross @Enselic: Do you use `FORMAT_VERSION`, and if not why? Would you like a fuller changelog, just a number, or nothing at all?

**Comment by notriddle** [comment]
> This could work, but will have a problem if two features are given the same name

Instead of names, pull request URLs?

**Comment by Urgau** [comment]
> Do you use `FORMAT_VERSION`, and if not why?

No, not yet; I just didn't found the time and the right way to get it but the solution from `smithy-rs` is quite simple so I've use it.

> Would you like a fuller changelog, just a number, or nothing at all?

Yes, a changelog would be useful but I don't think it's not required. The format number is good no need for more.
I would just like to have a recommended way to get and check it. The `smithy-rs` solution could maybe be integrated in the crate?
```rust
#[derive(Deserialize)]
struct CrateFormatVersion {
    format_version: u32,
}
```

**Comment by camelid** [comment]
> Instead of names, pull request URLs?

I think the issue with that is that you can't determine it before opening a PR, but we could have people just update that after opening their PR maybe?

**Comment by Enselic** [comment]
I currently do not make direct use of format version, but I think there is value in keeping it anyway.

There are several reasons I don't make direct use of it:
* It does not mean anything to regular users of my tools. If I say "you need format version 12" no one knows what that means. If I say "you need at least nightly-2022-03-04" people know what that means.
* In the past it JSON was sometimes changed without a bump in format version, which made nightly version the most accurate identifier anyway
* Sometimes it would be unnecessarily strict to enforce identical format version. For example, format version 12 output can be read by code that assume format version 11, because 12 only added fields/properties to existing structs. So requiring format version to be the same would be unnecessarily limiting.

That said, I think there is value in keeping format version, because it:
* Signals that the format version matters 
* Gives a quick indication of what kind of JSON you are dealing with if you are encountering some JSON without additional context

The risk of two features having the same logical name is low, but can be eliminated by scoping with username:

```
// Latest Feature: Foo by Enselic
```

I'm not a big fan of having to force push directly after creating a PR to update with the PR number

**Comment by jdisanti** [comment]
> Do you use FORMAT_VERSION, and if not why? Would you like a fuller changelog, just a number, or nothing at all?

I use `FORMAT_VERSION` to be notified of changes to the JSON format so that I can make sure my tool is working correctly with those changes. If you wanted to use a list of feature names, that works for me too since I could just compare the last known feature name with the end of that list.

Thinking about it more though, I don't absolutely need a format version or change log since I have a comprehensive test suite that should detect any breaking changes to the format that actually impact my tool.

**Comment by aDotInTheVoid** [comment]
This is at risk of happening again with #99688 and #99787 for the 16->17 change

**Comment by aDotInTheVoid** [comment]
This is at risk of happening with #127289 and #127276 for the 30->31 change. Both of these PR's are mine and were spawned from the same [zulip thread](https://rust-lang.zulipchat.com/#narrow/stream/266220-t-rustdoc/topic/.60ItemEnum.3A.3AOpaqueTy.60), which is why there's overlapping changes to such a low-traffic area of the codebase

**Comment by its-the-shrimp** [comment]
One solution to this could be releasing breaking changes to rustdoc-json-types only alongside a new compiler version itself, i.e. with the new 6-week cycle, if there is any change to be made, but such an approach would require stabilising the existence of JSON as an output format of rustdoc

**Comment by aDotInTheVoid** [comment]
This is at risk of happening again with #138109 and #137956 for the 40->41 change.

**Comment by aDotInTheVoid** [comment]
At risk with #139919 and #139393 for 43->44

**Comment by nnethercote** [comment]
> This could work, but will have a problem if two features are given the same name, and then we're back to where we were. If someone knows some git-trickery to get around that, that would be great, but if not, I still thing doing this (or something to this effect) would be worthwhile.

Three years later, no such git-trickery has been found. I think the comment is a fine idea that, while not foolproof, will greatly reduce the chance of this problem occurring. (I think "forgetting to update the comment" is much more likely than "two identical comments".)

So I've implemented this in #142601.
